### PR TITLE
Introduce `get_python_state_dict()` for Python 3.12 compatibility.

### PR DESCRIPTION
**Pull request by rwgk**
## Description
Based on PR #4293 but **without** ABI break.

Following a discussion under https://github.com/wjakob/nanobind/issues/96: to ensure compatibility with future Python versions, pybind11 needs to stop stashing its `internals` data in the global `builtins` dictionary.

To maximize backward compatibility, this PR preserves the status quo for `PYBIND11_INTERNALS_VERSION` `4` and Python versions <= 3.11. There is no benefit to changing anything for existing Python versions as long as `PYBIND11_INTERNALS_VERSION` `4` is used, except maybe for a very minor reduction in code complexity (a simpler preprocessor `#if` condition in `get_python_state_dict()`). That seems like an extremely small price to pay for maintaining backward compatibility.

Compatible with smart_holder (PR #4571) and pywrapcc (PR https://github.com/google/pywrapcc/pull/30007).

Tested with Python 3.12 via pywrapcc.

## Suggested changelog entry:

```rst

```

### CODE REVIEW

**Comment by Skylion007** [comment]
Hmm, thoughts on this @henryiii?

**Review by Skylion007 — APPROVED** [review]

**Comment by rwgk** [comment]
> Can we enable the CI for Python 3.12 yet, btw?

Yes, although not from sources. Each time there is a new alpha (or later beta) release we have to wait for a few days until `actions/setup-python@v4` pulls it in and we can test with it.

Currently I'm working with https://github.com/google/pywrapcc/blob/92741e75f85829d13ed18fbfa0fab491eeaa2896/.github/workflows/python312dev.yml

@henryiii didn't support that I wanted a separate yml for each pre-release Python version, and that I keep track of intermediate states in comments. I find it too difficult (and not very obvious) to repurpose upstream.yml, and believe that trying to keep something so fluid by nature pure is only slowing down progress, therefore I transferred the 3.12 testing to pywrapcc. If someone wants to jump in to maintain 3.12 testing here differently that would be great.

**Comment by rwgk** [comment]
Thanks @Skylion007!
I'll hold off merging until @henryiii has made the 2.10.4 release. (Not sure if merging now will make it more difficult to make the release.)

**Comment by rwgk** [comment]
Argh... I typed in the previous comment in the wrong window.

File: include/pybind11/detail/internals.h
Comment (Lalaland) [inline include/pybind11/detail/internals.h:426]: I think it might be simpler if you just had a define that you are always using internals version 5 if the python version is 3.12

**Review by Lalaland — APPROVED** [review]
Thanks for fixing this. It looks good, with one suggestion that you can make the ifdefs easier by bumping the abi version when 3.12 is detected

File: include/pybind11/detail/internals.h
Comment (rwgk) [reply include/pybind11/detail/internals.h]: I added this commit: a version bump for Python 3.12+ (I think of it as a free lunch version bump):

```diff
 #ifndef PYBIND11_INTERNALS_VERSION
-#    define PYBIND11_INTERNALS_VERSION 4
+#    if PY_VERSION_HEX >= 0x030C0000
+// Version bump for Python 3.12+, before first 3.12 beta release.
+#        define PYBIND11_INTERNALS_VERSION 5
+#    else
+#        define PYBIND11_INTERNALS_VERSION 4
+#    endif
 #endif
```

Then I looked long and hard at the condition here but eventually decided to leave it at a no-op shuffle, solely to increase readability slightly (see e8b34b473546c469f11cd2c2a0cb098ca12134e5).

> it might be simpler

Did you have in mind removing the `PY_VERSION_HEX < 0x030C0000` sub-condition here?

What worries me: if someone defines `PYBIND11_INTERNALS_VERSION 4` externally, we'd select the wrong code here if the sub-condition is missing. We don't know yet what the resulting errors might be, if any, for future Python releases.

I guess we could do something like this to guard the `PyEval_GetBuiltins()` path:

```
#if PY_VERSION_HEX >= 0x030C0000
#error Python 3.12+ requires PYBIND11_INTERNALS_VERSION >= 5
#endif
```

But that's strictly speaking an arbitrary forcing function. Maybe someone will have a good reason to experiment with `PYBIND11_INTERNALS_VERSION` 4 even for 3.12+? E.g. when debugging some issue with the TLS-related changes connected to version 5. This is totally made up of course, but I'm generally trying to stay clear of non-essential forcing functions (anecdotes of getting burned omitted).

Could you please re-approve if you think the latest version here is fine, or let me know if I misunderstood what you had in mind?

BTW: I re-tested with Python 3.12.0a6 under google/pywrapcc#30019 (https://github.com/google/pywrapcc/actions/runs/4526056859/jobs/7970965421?pr=30019).

Comment (Lalaland) [reply include/pybind11/detail/internals.h]: > But that's strictly speaking an arbitrary forcing function.

Yeah, it's an "arbitrary forcing function", but it reduces the matrix of Python version / pybind11 version abi combinations you have to support. I don't understand why you would want version abi version 4 and python 3.12 together, but if you do, feel free to disregard this suggestion because I was assuming that combination was sorta worthless (since you can't get abi version 4 and python 3.11 code to work together anyways).

Your code in mind is exactly what I had in mind, but here are a couple of suggestions.

```
#if PY_VERSION_HEX < 0x03080000                                                                   \
    || (PYBIND11_INTERNALS_VERSION <= 4 && PY_VERSION_HEX < 0x030C0000) || defined(PYPY_VERSION)
    state_dict = reinterpret_borrow<object>(PyEval_GetBuiltins());
#else
```

Why isn't this just

```
#if PYBIND11_INTERNALS_VERSION <= 4
    state_dict = reinterpret_borrow<object>(PyEval_GetBuiltins());
#else
```

> I guess we could do something like this to guard the PyEval_GetBuiltins() path:

I would add this to the very top where versions are defined
```

#ifndef PYBIND11_INTERNALS_VERSION
#    if PY_VERSION_HEX >= 0x030C0000
// Version bump for Python 3.12+, before first 3.12 beta release.
#        define PYBIND11_INTERNALS_VERSION 5
#    else
#        define PYBIND11_INTERNALS_VERSION 4
#    endif
#endif

// Require at least version 5 ABI for Python 3.12+
static_assert(!(PY_VERSION_HEX == 0x030C0000) || (PYBIND11_INTERNALS_VERSION >= 5);

```

Comment (rwgk) [reply include/pybind11/detail/internals.h]: > Why isn't this just
> 
> ```
> #if PYBIND11_INTERNALS_VERSION <= 4
> ```

We need to keep the < Python 3.8 condition because `_PyInterpreterState_Get()` only appeared with Python 3.8 (I looked in the 3.6 and 3.7 sources yesterday).

I'm trying the simpler condition right now. I'm not sure what we need to do for PyPy.

I'm on board enforcing ABI version 5 for 3.12+ (see latest code here), but just to make sure I'm not missing something:

> (since you can't get abi version 4 and python 3.11 code to work together anyways)

Did you mean to write 3.12 here?  —  We are testing with 3.11, e.g.:

```
============================= test session starts ==============================
platform linux -- Python 3.11.2, pytest-7.0.0, pluggy-1.0.0
C++ Info: 9.4.0 C++11 __pybind11_internals_v4_gcc_libstdcpp_cxxabi1013__ PYBIND11_SIMPLE_GIL_MANAGEMENT=True
```

Currently ABI 4 & 3.12 still work (from [pywrapcc GHA](https://github.com/google/pywrapcc/actions/runs/4461749631/jobs/7835824105)):

```
============================= test session starts ==============================
platform linux -- Python 3.12.0a6, pytest-7.0.0, pluggy-1.0.0
C++ Info: 11.3.0 C++11 __pybind11_internals_v4_gcc_libstdcpp_cxxabi1016_sh_def__ PYBIND11_SIMPLE_GIL_MANAGEMENT=False
```

Comment (Lalaland) [reply include/pybind11/detail/internals.h]: > (since you can't get abi version 4 and python 3.11 code to work together anyways)

Oops, typo. Yep, Meant that you can't get abi version 4 Python 3.12 code and abi version 4 Python 3.11 code to work together

> I'm not sure what we need to do for PyPy.

PyPy should be following the CPython API, so we should be able to use identical code for both it and CPython. Unless you know otherwise of course.

> We need to keep the < Python 3.8 condition because _PyInterpreterState_Get() only appeared with Python 3.8 (I looked in the 3.6 and 3.7 sources yesterday).

I would add a wrapper for _PyInterpreterState_Get to pybind11/detail/common.h with #ifdefs to switch back and forth between the _ and regular version depending on the Python version.

Comment (rwgk) [reply include/pybind11/detail/internals.h]: > PyPy should be following the CPython API, so we should be able to use identical code for both it and CPython.

Unfortunately not. PyPy with 3.8 and 3.9 still don't work (our GHA). I just looked in the [latest PyPy sources](https://foss.heptapod.net/pypy/pypy): `PyInterpreterState_Get` still doesn't appear there. I tried to use `PY_VERSION_HEX` but apparently that's not meaningful with PyPy. It's unclear to me at the moment if/how we can do better than just testing for `PYPY_VERSION`. Leaving that for later. While looking around I saw that there are 7.3.10 and 7.3.11 releases which we don't have in our GHA. Something to do separately maybe.

> I would add a wrapper for _PyInterpreterState_Get to pybind11/detail/common.h with #ifdefs to switch back and forth between the _ and regular version depending on the Python version.

This `get_python_state_dict()` is essentially that already. I don't feel factoring out the branch for >= 3.9 will be helpful, especially given the PyPy complication.

Comment (Lalaland) [reply include/pybind11/detail/internals.h]: Arg, another unfortunate PyPy incompatibility. I guess this is the best we can do.

**Review by Lalaland — APPROVED** [review]

**Comment by rwgk** [comment]
I retested with Python 3.12a6 under google/pywrapcc#30019: all good there, too.

Thanks @Lalaland and @Skylion007 for the reviews!

**Pull request merged**
